metrics/prometheus_machine.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	info "github.com/google/cadvisor/info/v1"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"k8s.io/klog"
+)
+
+var baseLabelsNames = []string{"machine_id", "system_uuid", "boot_id"}
+
+const (
+	prometheusModeLabelName = "mode"
+	prometheusTypeLabelName = "type"
+
+	nvmMemoryMode    = "memory_mode"
+	nvmAppDirectMode = "app_direct_mode"
+
+	memoryByTypeDimmCountKey    = "DimmCount"
+	memoryByTypeDimmCapacityKey = "Capacity"
+	memoryByTypeAllType         = "all"
```

Comment:
IIUC, the "all" type is the sum of all other types, right?  If this is the case, we should probably omit it, since it is a best practice for metrics to be able to be aggregated by label.  E.g. if we have 
```
machine_dimm_count{"type": "foo"} 1
machine_dimm_count{"type": "bar"} 2
```
We can query `sum(machine_dimm_count)` to get the total, 3.  Or we can choose to do `sum by type (machine_dimm_count)` to get our breakdown back.

That should simplify the code below considerably

Reply:
Now from this information available on API:
```
"memory_by_type": {
    "Non-volatile-RAM": {
      "capacity": 2168421613568,
      "dimm_count": 8
    },
    "Unbuffered-DDR4": {
      "capacity": 412316860416,
      "dimm_count": 12
    }
  },
```
following metrics are created:
```
# HELP machine_dimm_capacity_bytes Total RAM DIMM capacity (all types memory modules) value labeled by dimm type.
# TYPE machine_dimm_capacity_bytes gauge
machine_dimm_capacity_bytes{boot_id="2cd60c6c-ed61-4ccb-a3c5-98de349eb8f5",machine_id="39efbe7cf4ee42bcb6288a810d2f3ace",system_uuid="80164458-B7A3-E811-906E-00163566263E",type="Non-volatile-RAM"} 2.168421613568e+12
machine_dimm_capacity_bytes{boot_id="2cd60c6c-ed61-4ccb-a3c5-98de349eb8f5",machine_id="39efbe7cf4ee42bcb6288a810d2f3ace",system_uuid="80164458-B7A3-E811-906E-00163566263E",type="Unbuffered-DDR4"} 4.12316860416e+11
# HELP machine_dimm_count Number of RAM DIMM (all types memory modules) value labeled by dimm type.
# TYPE machine_dimm_count gauge
machine_dimm_count{boot_id="2cd60c6c-ed61-4ccb-a3c5-98de349eb8f5",machine_id="39efbe7cf4ee42bcb6288a810d2f3ace",system_uuid="80164458-B7A3-E811-906E-00163566263E",type="Non-volatile-RAM"} 8
machine_dimm_count{boot_id="2cd60c6c-ed61-4ccb-a3c5-98de349eb8f5",machine_id="39efbe7cf4ee42bcb6288a810d2f3ace",system_uuid="80164458-B7A3-E811-906E-00163566263E",type="Unbuffered-DDR4"} 12
```